### PR TITLE
Prism/create gem fix toast notification and gem catalog refresh

### PR DESCRIPTION
**Pull request by tkothadev**
## What does this PR do?

Recently @AMZN-alexpete and I discovered major usability issues with the Create a gem workflow. One of those issues was the severe lack of notification upon successful completion of gem creation.

Here was the behavior of the application prior to changes proposed in this PR:
CreateAGemWorkflowRun

Although I was able to complete the gem creation workflow, it was far from satisfactory. It went to a blank repo screen, and even the gem catalog did not update with the newly made gem (as if it did not exist).

I was able to confirm however that the gem was generated at my desired directory, meaning the actual creation of the gem was successful, but the Project Manager seemed unaware of those changes.

This is the summary of what I discovered to be the issue with the current code:

After creation of the gem, a signal for `CreateButtonPressed` is emitted. The respective `ScreenWidget` housing the `GemCatalogScreen` at the time (`UpdateProjectCtrl` at the time of this PR), processes this via a slot, which is only mapped to the standard `HandleBackButton` function, which does not have enough logic to actually handle gem creation specific details. 

This leads to the following issues:

1. No notifications are sent to the catalog screen
2. `HandleBackButton` actually messes up the desired ordering of screen transitions. A stack based transition is not appropriate, because we need to skip over the repo screen to go back to the main catalog page.
3. The gem model for the catalog is not updated and persisted.

In other words, `CreateAGem`  uses python bindings to actually create the gem, but doesnt update the model. When compared to functionality for adding an existing gem, this difference can be plainly seen

## How was this PR tested?

Testing was performed manually to verify behaviors were correct.

These are the results of the new changes. 

SuccessfulCreateAGemNotification

As shown in the gif, the Create a Gem workflow shows improved behavior with notifying the user of success, and it keeps changes permanent in the catalog for future access to the gem.

### CODE REVIEW

File: Code/Tools/ProjectManager/Source/CreateAGemScreen.cpp [hunk 1]
```diff
@@ -400,7 +400,16 @@ namespace O3DE::ProjectManager
                 auto result = PythonBindingsInterface::Get()->CreateGem(templateLocation, m_createGemInfo);
                 if (result.IsSuccess())
                 {
-                    emit CreateButtonPressed();
+                    //make sure the gem is registered, otherwise our changes will not persist
+                    auto registerResult = PythonBindingsInterface::Get()->RegisterGem(m_createGemInfo.m_path);
```
Comment (AMZN-Phil) [inline Code/Tools/ProjectManager/Source/CreateAGemScreen.cpp:404]: CreateGem can register the gem after creating it which would save the second call to PythonBindings/Python. It looks like the registerGem parameter of CreateGem is used as no_register right now though so it's likely doing the opposite of what it should

File: Code/Tools/ProjectManager/Source/CreateAGemScreen.h [hunk 2]
```diff
@@ -35,7 +35,7 @@ namespace O3DE::ProjectManager
         ~CreateGem() = default;
 
     signals:
-        void CreateButtonPressed();
+        void CreateButtonPressed(GemInfo gemInfo);
```
Comment (AMZN-alexpete) [inline Code/Tools/ProjectManager/Source/CreateAGemScreen.h:38]: Lets use a const ref to avoid copying the data more than needed
```suggestion
        void CreateButtonPressed(const GemInfo& gemInfo);
```

File: Code/Tools/ProjectManager/Source/GemCatalog/GemCatalogScreen.h [hunk 3]
```diff
@@ -44,6 +44,10 @@ namespace O3DE::ProjectManager
         bool IsTab() override;
         void NotifyCurrentScreen() override;
 
+        void AddToGemModel(GemInfo gemInfo);
+
+        void ShowStandardToastNotification(QString notification);
```
Comment (AMZN-alexpete) [inline Code/Tools/ProjectManager/Source/GemCatalog/GemCatalogScreen.h:49]: both of these should use const ref to avoid unnecessary copying of the parameter
```suggestion
        void AddToGemModel(const GemInfo& gemInfo);

        void ShowStandardToastNotification(const QString& notification);
```

**Comment by amzn-leenguy** [comment]
UX pov: It was strange to me that there was 2 footer bars with 4 CTA buttons. I understand that the last footer row was for the previous workflow, but ideally i would like to offer the user only one set of footer CTA buttons. Which is the one of the workflow that you selected (to create a gem). having double is strange and confusing.

In regards to the final screen it should go to after you have created a gem, after editing a project, I feel like it should take the user back to where they came from. Which in your gif example, it should take them back to the gem catalog with a toast message notifying the user that their gem creation was successful. And it should appear in the list.

Please let me know if you are able to call the toast messaging system in the gem catalog. the same one that appears when you add enable gems with the toggles in the regular gem catalog UX workflow.

**Comment by tkothadev** [comment]
> UX pov: It was strange to me that there was 2 footer bars with 4 CTA buttons. I understand that the last footer row was for the previous workflow, but ideally i would like to offer the user only one set of footer CTA buttons. Which is the one of the workflow that you selected (to create a gem). having double is strange and confusing.

You are correct, this is in fact a bug, but it’s a separate one that was out of scope for the current PR. It already existed prior to my starting work on the fixes. I wanted to dedicate a separate PR to address that issue properly. This will occur later on after stabilization next week.

> In regards to the final screen it should go to after you have created a gem, after editing a project, I feel like it should take the user back to where they came from. Which in your gif example, it should take them back to the gem catalog with a toast message notifying the user that their gem creation was successful. And it should appear in the list.
> 
> Please let me know if you are able to call the toast messaging system in the gem catalog. the same one that appears when you add enable gems with the toggles in the regular gem catalog UX workflow.

Yes this is in fact the case. The changes I added to this PR address what is requested here, and I am able to use the existing toast messaging system in the gem catalog. I posted the gif showing off these features originally in the first PR comment, but I'll repost the gif here again for clarity:

File: Code/Tools/ProjectManager/Source/PythonBindings.cpp [hunk 4]
```diff
@@ -852,7 +852,7 @@ namespace O3DE::ProjectManager
                     "icon_path"_a = QString_To_Py_Path(gemInfo.m_iconPath),
                     "documentation_url"_a = QString_To_Py_String(gemInfo.m_documentationLink),
                     "repo_uri"_a = QString_To_Py_String(gemInfo.m_repoUri),
-                    "no_register"_a = registerGem)
+                    "no_register"_a = !registerGem)
```
Comment (AMZN-alexpete) [inline Code/Tools/ProjectManager/Source/PythonBindings.cpp:855]: nice!

**Review by AMZN-Phil — APPROVED** [review]

File: Code/Tools/ProjectManager/Source/GemCatalog/GemCatalogScreen.h [hunk 5]
```diff
@@ -12,7 +12,7 @@
 #include <ScreenWidget.h>
 #include <AzCore/std/smart_ptr/unique_ptr.h>
 #include <AzToolsFramework/UI/Notifications/ToastNotificationsView.h>
-
+#include "GemInfo.h"
```
Comment (AMZN-alexpete) [inline Code/Tools/ProjectManager/Source/GemCatalog/GemCatalogScreen.h:15]: though quotes are more in line with most coding standards and [what quote vs angle bracket denotes](https://gcc.gnu.org/onlinedocs/cpp/Include-Syntax.html), [our own coding guidelines use angle brackets](https://github.com/o3de/sig-core/blob/main/governance/Coding-Standards-and-Style-Guide.md#including-headers) and that will be more consistent with the rest of the code in this project.
```suggestion
#include <GemInfo.h>
```

**Review by AMZN-alexpete — APPROVED** [review]
Just one minor suggestion on syntax for the GemInfo.h include

**Comment by amzn-leenguy** [comment]
Great work @tkothadev and thanks for the reply. Could we update the toast label to:

For new gems created:
<Name> Gem has been created.

For editing a gem and updating it (later on):
<Name> gem has been updated.

No need to add successfully, as it may add unnecessary character length and if the gem name is long, we may have long toasts.

**Comment by tkothadev** [comment]
@amzn-leenguy 
A question for clarification:
1) Is it fine if we retain the gem name in the toast notification? For example, if we tweak the message to:
Gem created: <GEM_NAME>

Reason I ask is because it might be reassuring to see the name in the notification, as a way of communicating that the gem really was registered in the system.

**Comment by amzn-leenguy** [comment]
Sorry I added Gem_NAME but it seems GITHUB removed my words .... brb editing my post.

**Comment by amzn-leenguy** [comment]
> @amzn-leenguy A question for clarification: Is it fine if we retain the gem name in the toast notification? For example, if we tweak the message to: Gem created: <GEM_NAME>
> 
> Reason I ask is because it might be reassuring to see the name in the notification, as a way of communicating that the gem really was registered in the system.

I've updated my copy and removed the brackets so git hub comments doesn't remove my text suggestions.

**Comment by tkothadev** [comment]
Thanks. I'll add the change in. To confirm, I will be making a single change such that the toast notification renders the following text:

GEM_Name Gem has been created.

And follow that up with a screenshot showing just that.

**Comment by amzn-leenguy** [comment]
> Thanks. I'll add the change in. To confirm, I will be making a single change such that the toast notification renders the following text:
> 
> GEM_Name Gem has been created.
> 
> And follow that up with a screenshot showing just that.

This is correct. Thanks!

**Comment by tkothadev** [comment]
@amzn-leenguy 

Here's the screenshot for your suggested change:
image

If this is good, I will commit it to this PR.

**Review by amzn-leenguy — APPROVED** [review]
